Enhance docs with GitBook (#398)
src-d/lookout: docs/analyzers-creation.md

> # Implementing Your Own Analyzer

Reviewer: I wish we had links to examples somewhere in this document. Because after reading it I still have no idea how to start.

Author: Was it improved with the provided examples?
eb222fb

Reviewer: yes, definitely! Though python example is incorrect (it exits after 60 seconds).

Author: I think it's up to the user to decide how to proceed to wait before stopping the server.

I also tried to be concise, so I preferred not to handle that in detail.
On the other hand, it is also explained in the gRPC.io docs about how to start the server, already linked in our docs just above the python example.

_Please refer to the [**official Protocol Buffers** documentation](https://developers.google.com/protocol-buffers/) to learn how to get started with Protocol Buffers._

To implement your own analyzer you only need to create a gRPC service implementing [Analyzer service](https://github.com/src-d/lookout-sdk/blob/master/proto/lookout/sdk/service_analyzer.proto#L30) interface:

```protobuf
service Analyzer {
  rpc NotifyReviewEvent (ReviewEvent) returns (EventResponse);
  rpc NotifyPushEvent (PushEvent) returns (EventResponse);
}
```

You can create a new analyzer in any language that supports protocol buffers, generating code from [the `.proto` definitions](https://github.com/src-d/lookout-sdk/tree/master/proto/lookout/sdk). The resulting code will provide data access classes, with accessors for each field, as well as methods to serialize/parse the message structures to/from bytes.

## Caveats

All the analyzers should consider [the caveats described by the SDK](https://github.com/src-d/lookout-sdk#caveats).

## Fetching Changes, UASTs or Languages from lookout DataService

> The analyzer should not need to take care of managing git data, nor storing the repositories, nor guessing the programming language used in the files nor parsing its UASTs; to do so, it should request **lookout DataService** for such data.

Reviewer 2: It sounds a bit convoluted to me. What about
```suggestion
Lookout will take care of dealing with Git repositories, UAST extraction, programming language detection, etc. Your analyzer will be able to use the **lookout DataService** to query all this data.
```

> You can read more about it in [**lookout DataService** section](architecture.md#lookout-dataservice).

Reviewer 2: ```suggestion
You can read more about it in the [**lookout DataService** section](architecture.md#lookout-dataservice).
```

> ## How to Locally Try an Analyzer

Reviewer 2: ```suggestion
## How to Test an Analyzer Locally
```

_Please refer to [**lookout-sdk** docs](lookout-sdk.md) to see how to locally test an analyzer without accessing GitHub at all._

# Using Pregenerated Code from the SDK

> If you're creating the analyzer in Golang or Python, you'll find pre-generated libraries in the [lookout-sdk repository](https://github.com/src-d/lookout-sdk) so you'll have to generate nothing.

Reviewer 2: I would add something about the helpers. What about:

```suggestion
If you're creating your analyzer in Golang or Python, you'll find pre-generated libraries in the [lookout-sdk repository](https://github.com/src-d/lookout-sdk). The SDK libraries also come with helpers to deal with gRPC caveats.
```

You can do as it follows:

## Golang

> Import and use `github.com/src-d/lookout-sdk/pb`.

Reviewer 2: ```suggestion
Import and use `gopkg.in/src-d/lookout-sdk.v0/pb`.
```

> The analyzer must implement the [AnalyzerClient interface](https://github.com/src-d/lookout-sdk/blob/master/pb/service_analyzer.pb.go#L129):

Reviewer 2: This should be AnalyzerServer, not the client, right?
Also, maybe it's better to link to godocs.

```suggestion
The analyzer must implement the [AnalyzerServer interface](https://godoc.org/gopkg.in/src-d/lookout-sdk.v0/pb#AnalyzerServer):
```

> ```go

Reviewer 2: I propose this other code:

```go
import "gopkg.in/src-d/lookout-sdk.v0/pb"

type Analyzer struct {}

func (a *Analyzer) NotifyReviewEvent(ctx context.Context, e *lookout.ReviewEvent) (
...
}

func (a *Analyzer) NotifyPushEvent(ctx context.Context, e *lookout.PushEvent) (*lookout.EventResponse, error) {
...
}

Reviewer: 👍 current code is invalid

Reviewer 2: proposed*

And yes, LGTM :+1:

import github.com/src-d/lookout-sdk/pb
type AnalyzerClient interface {
  NotifyReviewEvent(context.Context, *pb.ReviewEvent, opts ...grpc.CallOption) (*pb.EventResponse, error)
  NotifyPushEvent(context.Context, *pb.PushEvent, opts ...grpc.CallOption) (*pb.EventResponse, error)
}
```

> Once the gRPC server is run, it will listen for requests from `lookoutd`.

Reviewer 2: ```suggestion
Once the gRPC server is started, it will listen for requests from `lookoutd`.
```

## Python

Install and use [`lookout_sdk`](https://pypi.org/project/lookout-sdk) python library:

```shell
$ pip install lookout-sdk
```

> The analyzer class will extend [AnalyzerServicer](https://github.com/src-d/lookout-sdk/blob/master/python/lookout/sdk/service_analyzer_pb2_grpc.py#34):

Reviewer 2: Bad link,
```suggestion
The analyzer class will extend [AnalyzerServicer](https://github.com/src-d/lookout-sdk/blob/master/python/lookout/sdk/service_analyzer_pb2_grpc.py#L34):
```

Reviewer: in general it's bad idea to link on line on master. It will become wrong sooner or later. Better link to the last tag. And just grep when we release new version.

```python

> class AnalyzerServicer(object):

Reviewer: if we change go example above we need to change python too
